docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,195 @@
+
+Once found, CAPI will create and connect MachinePool Machines to each resource. A MachinePool Machine is implemented as a Cluster API Machine that is owned by a MachinePool. If the field is empty, no MachinePool Machines will be created, and the overall behavior is the same as before this proposal was implemented.
+
+It is the responsibility of each provider to populate `InfrastructureMachineSelector` and to create provider-specific MachinePool Machine resources behind each Machine. For example, the Docker provider will create a DockerMachinePoolMachine to represent the container instance behind the Machine in the infrastructure provider's MachinePool. It will also ensure that the DockerMachinePoolMachine is labeled such that the `InfrastructureMachineSelector` can be used to find it.
+
+![MachinePool Controller Infrastructure Reference Changed Sequence](images/machinepool-machines/infraref-changed-sequence.png)
+
+When a MachinePool Machine is deleted manually, the system will delete the corresponding provider-specific resource. The opposite is also true: when a provider-specific resource is deleted, the system will delete the corresponding MachinePool Machine. This happens by virtue of the infrastructureRef <-> ownerRef relationship.
+
+In both cases, the MachinePool will notice the missing replica and create a new one in order to maintain the desired number of replicas. To scale down by removing a specific instance, that Machine should be given the "cluster.x-k8s.io/delete-machine" annotation and then the replicaCount on the MachinePool should be decremented.
+
+### User Stories
+
+--------------------------------------------
+| ID | Story                               |
+|----|-------------------------------------|
+| U1 | Cluster Autoscaler                  |
+| U2 | MachinePool Machine Remediation     |
+| U3 | MachinePool Machine Rolling Updates |
+--------------------------------------------
+
+#### Story U1
+
+A cluster admin is tasked with using Cluster Autoscaler to scale a MachinePool as required by machine load. The MachinePool should provide a mechanism for Cluster Autoscaler to increase replica count, and provide a mechanism for Cluster Autoscaler to reduce replica count by selecting an optimal machine for deletion.
+
+#### Story U2
+
+A cluster admin would like to configure CAPI to remediate unhealthy MachinePool Machines to ensure a healthy fleet of Machines.
+
+#### Story U3
+
+A cluster admin updates a MachinePool to a newer Kubernetes version and would like to configure the strategy for that deployment so that the MachinePool will progressively roll out the new version of the machines. They would like this operation to cordon and drain each node to minimize workload disruptions.
```

Comment:
Again I somehow dont understand which component will own this? Will CAPI machinepool controller do this?

Reply:
In this case the CAPI machinepool controller's existing logic should own everything. Behind this point is the idea that CAPZ has already implemented cordon + drain separately for its own MachinePool Machine implementation, but should be able to remove some of that duplicated logic now.

@shyamradhakrishnan I apologize for missing these comments for so long! Let me know what other questions you have.